transformer_engine/pytorch/float8_tensor.py

```diff
@@ -550,44 +544,71 @@ def __torch_dispatch__(cls, func, types, args, kwargs=None):
-            src = src.expand(dst.size())
-            src = src.to(
-                device=dst.device,
-                memory_format=torch.contiguous_format,
-            )
+                raise RuntimeError(
+                    "Attempted to copy from something that isn't a PyTorch tensor"
+                )
 
-            # Update scaling factor if FP8 meta tensors are available
-            if dst._fp8_meta is None:
-                scale = dst._scale_inv.reciprocal()
-            else:
-                fp8_meta_key = FP8GlobalStateManager.get_meta_tensor_key(
-                    forward=dst._fp8_meta_forward,
+            # Special handling based on which tensors are FP8
+            dst_is_fp8 = isinstance(dst, Float8Tensor)
+            src_is_fp8 = isinstance(src, Float8Tensor)
+            if dst_is_fp8 and src_is_fp8:
+
+                # Directly copy FP8 data if possible
+                if dst._fp8_dtype == src._fp8_dtype:
+                    dst._data.copy_(src._data)
+                    dst._scale_inv = src._scale_inv.clone()
+                else:
+                    dst.copy_(src.from_float8())
+
+            elif not dst_is_fp8 and src_is_fp8:
+
+                # Cast source tensor to higher precision
+                dst.copy_(src.from_float8())
+
+            elif dst_is_fp8 and not src_is_fp8:
+
+                # Make sure input is in expected format
+                src = src.expand(dst.size())
+                src = src.to(
+                    device=dst.device,
+                    memory_format=torch.contiguous_format,
                 )
-                scale = dst._fp8_meta[fp8_meta_key].scale[dst._fp8_meta_index]
-                dst._scale_inv = scale.detach().view(1).reciprocal()
 
-            # Cast to FP8
-            tex.cast_to_fp8_noalloc(
-                src.view(1,-1),
-                scale,
-                dst._data.view(1,-1),
-                torch.empty_like(dst._scale_inv),  # amax
-                dst._scale_inv,
-                dst._fp8_dtype,
-            )
+                # Update scaling factor if FP8 meta tensors are available
+                if dst._fp8_meta is None:
+                    scale = dst._scale_inv.reciprocal()
+                else:
+                    fp8_meta_key = FP8GlobalStateManager.get_meta_tensor_key(
+                        forward=dst._fp8_meta_forward,
+                    )
+                    scale = dst._fp8_meta[fp8_meta_key].scale[dst._fp8_meta_index]
+                    dst._scale_inv = scale.detach().view(1).reciprocal()
+
+                # Cast to FP8
+                if not dst._data.is_contiguous():
+                    raise RuntimeError("Transformer Engine cast kernels require contiguous data")
+                tex.cast_to_fp8_noalloc(
+                    src.view(1,-1),
+                    scale,
+                    dst._data.view(1,-1),
+                    torch.empty_like(dst._scale_inv),  # amax
```

Comment:
Maybe just need more clarity - Is the `amax` being updated elsewhere? I guess in dist opt but for other optimizers, this needs to be handled carefully?

Reply:
This amax isn't used anywhere. It's included in the kernel since it's used for the FP8 recipe when casting weights and activations. It might work if I pass in an empty tensor - the kernel does check if amax is `nullptr` - but we would need to check.